server/api.js

```diff
@@ -92,4 +108,56 @@ router.post("/alternatives", async (req, res) => {
 	}
 });
 
+router.post("/saveUserData", async (req, res) => {
+	try {
+		const { name, email, sub, input, output, alternative } = req.body;
+
+		// Check if the user already exists in the database
+		const existingUser = await db.query(
+			"SELECT * FROM users WHERE email = $1",
+			[email]
+		);
+
+		if (existingUser.rows.length > 0) {
+			// User already exists in the database
+
+			// Check if the input and output values exist in the history table
+			const existingHistory = await db.query(
+				"SELECT * FROM history WHERE input = $1 AND output = $2",
+				[input, output]
+			);
+
+			if (existingHistory.rows.length > 0) {
+				// Input and output values exist in the history table, perform an update
+				await db.query(
+					"UPDATE history SET alternative = $1 WHERE input = $2 AND output = $3",
+					[alternative, input, output]
+				);
+			} else {
+				// Input and output values don't exist in the history table, insert a new row
+				await db.query(
+					"INSERT INTO history (email, input, output, alternative) VALUES ($1, $2, $3, $4)",
+					[email, input, output, alternative]
+				);
+			}
+		} else {
+			// User not found, save the user data in the database
+			await db.query(
+				"INSERT INTO users (name, email, google_id) VALUES ($1, $2, $3)",
+				[name, email, sub]
+			);
+
+			// Insert a new row in the history table
+			await db.query(
+				"INSERT INTO history (email, input, output, alternative) VALUES ($1, $2, $3, $4)",
+				[email, input, output, alternative]
+			);
+		}
```

Comment:
duplication here, can these be broken out into a separate function?

Reply:
Done.